network/local.go

```diff
@@ -0,0 +1,459 @@
+func LocalReset() {
+	defaultLocalContext = NewLocalContext()
+
+}
+
+// IsListening returns true if the remote address is listening "virtually"
+func (ccc *LocalContext) IsListening(remote Address) bool {
+	ccc.Lock()
+	defer ccc.Unlock()
+	_, ok := ccc.listening[remote]
+	return ok
+}
+
+// Listening put the address as "listening" mode. If a user connects to this
+// addr, this function will be called.
+func (ccc *LocalContext) Listening(addr Address, fn func(Conn)) {
+	ccc.Lock()
+	defer ccc.Unlock()
+	ccc.listening[addr] = fn
+}
+
+// StopListening remove the address from the "listening" mode
+func (ccc *LocalContext) StopListening(addr Address) {
+	ccc.Lock()
+	defer ccc.Unlock()
+	delete(ccc.listening, addr)
+}
+
+// Connect will check if the remote address is listening, if yes it creates
+// the two connections, and launch the listening function in a go routine.
+// It returns the outgoing connection with any error.
+func (ccc *LocalContext) Connect(local, remote Address) (*LocalConn, error) {
+	ccc.Lock()
+	defer ccc.Unlock()
+
+	fn, ok := ccc.listening[remote]
+	if !ok {
+		return nil, fmt.Errorf("%s can't connect to %s: it's not listening", local, remote)
+	}
+
+	outEndpoint := endpoint{local, ccc.baseUID}
+	ccc.baseUID++
+
+	incEndpoint := endpoint{remote, ccc.baseUID}
+	ccc.baseUID++
+
+	outgoing := newLocalConn(ccc, outEndpoint, incEndpoint)
+	incoming := newLocalConn(ccc, incEndpoint, outEndpoint)
+
+	// outgoing knows how to store packet into the incoming's queue
+	ccc.queues[outEndpoint] = outgoing.connQueue
+	// incoming knows how to store packet into the outgoing's queue
+	ccc.queues[incEndpoint] = incoming.connQueue
+
+	go fn(incoming)
+	return outgoing, nil
+}
+
+// Send will get the connection denoted by this endpoint and will call queueMsg
+// with the packet as argument on it. It returns ErrClosed if it does not find
+// the connection.
+func (ccc *LocalContext) Send(e endpoint, nm Packet) error {
+	ccc.Lock()
+	defer ccc.Unlock()
+	q, ok := ccc.queues[e]
+	if !ok {
+		return ErrClosed
+	}
+
+	q.Push(nm)
+	return nil
+}
+
+// Close will get the connection denoted by this endpoint and will Close it if
+// present.
+func (ccc *LocalContext) Close(conn *LocalConn) {
+	ccc.Lock()
+	defer ccc.Unlock()
+	// delete this conn
+	delete(ccc.queues, conn.local)
+	// and delete the remote one + close it
+	remote, ok := ccc.queues[conn.remote]
+	if !ok {
+		return
+	}
+	delete(ccc.queues, conn.remote)
+	remote.Close()
+}
+
+// Len returns how many local connections is there
+func (ccc *LocalContext) Len() int {
+	ccc.Lock()
+	defer ccc.Unlock()
+	return len(ccc.queues)
+}
+
+// LocalConn is a connection that send and receive messages through channels
+type LocalConn struct {
+	local  endpoint
+	remote endpoint
+
+	// connQueue is the part that is accesible from the LocalContext (i.e. is
+	// shared). Reason why we can't directly share LocalConn is because go test
+	// -race detects as data race (while it's *protected*)
```

Comment:
I'm not sure it's a bug. 
`LocalContext` has two connections `c1` and `c2`. `c1` is `Send()`ing to `c2` in one goroutine(g1), and *at the same time* , `c2` is receiving from `c1` in another goroutine (g2). In g1, `c1` will call `localcontext.Send()` which will call `c2.Push`. So g1 is *modifying* `c2`'s BUFFER, and  g2 is *reading* `c2`'s BUFFER. Even though all operations to the buffer are protected by mutexes, the *access to the connections's buffer* is not (g1 access `c2` and g2 too). 
So I'm not sure it's a bug or if I do strange things.